File: Design-Patterns-Using-Cpp/ArchitecturalPatterns/MVC.cpp
#include <iostream>
#include <vector>
#include <string>
#include <memory>
#include <sstream>
#include <algorithm>


// Observer interface
class Observer {
public:
    virtual void update() = 0;
};


// Model
class Model {
private:
    std::string data;
    std::vector<Observer*> observers;

public:
    void setData(std::string d) {
        std::cout << "Model : Set data." << std::endl;
        data = d;
        notifyObservers();
    }

    std::string getData() {
        std::cout << "Model : Get data." << std::endl;
        return data;
    }

    void addObserver(Observer* observer) {
        observers.push_back(observer);
    }

    void removeObserver(Observer* observer) {
        observers.erase(std::remove(observers.begin(), observers.end(), observer), observers.end());
    }

    void notifyObservers() {
        std::cout << "Model : Notify observers." << std::endl;
        for (auto observer : observers) {
            observer->update();
        }
    }
};



// View
class View  : public Observer {
private:
    class Controller* controller;
    Model* model;

public:
    View(Model* model, Controller* controller) : model(model), controller(controller) {
        model->addObserver(this);
    }

    void update() {
        std::cout << "View : Update." << std::endl;
        std::string data = model->getData();
        std::cout << "Data: " << data << std::endl;
    }

    void getUserInput() ;
};

// Controller
class Controller {
private:
    Model* model;
    View* view;

public:
    Controller(Model* model) : model(model) {}

    void handleUserInput(std::string userInput) {
        std::cout << "Controller : Handle user input." << std::endl;
        model->setData(userInput);
    }

    void setView(View* v) {
        this->view = v;
    }

    void update() {
        view->update();
    }
};

void View::getUserInput() 
{
    std::cout << "View : Enter user input: ";
    //std::string userInput;
    std::string userInput = "Hello, World!";
    std::cout << userInput << std::endl;
    // std::getline(std::cin, userInput); // Uncomment for user input    
    controller->handleUserInput(userInput);
}

// Client code
int main() {
    Model model;
    Controller controller(&model);
    View view(&model, &controller);
    controller.setView(&view);
    view.getUserInput();
    return 0;
}

/*
View : Enter user input: Hello, World!
Controller : Handle user input.
Model : Set data.
Model : Notify observers.
View : Update.
Model : Get data.
Data: Hello, World!
*/

File: Design-Patterns-Using-Swift/ArchitecturalPatterns/MVC.swift
import Foundation

// Model
class Model {
    private var data: String = ""
    private var observers: [View] = []

    func setData(_ data: String) {
        print("Model : Set data.")
        self.data = data
        notifyObservers()
    }

    func getData() -> String {
        print("Model : Get data.")
        return data
    }

    func addObserver(_ observer: View) {
        observers.append(observer)
    }

    func removeObserver(_ observer: View) {
        if let index = observers.firstIndex(where: { $0 === observer }) {
            observers.remove(at: index)
        }
    }

    private func notifyObservers() {
        print("Model : Notify observers.")
        for observer in observers {
            observer.update()
        }
    }
}

// View
class View {
    private var controller: Controller
    private var model: Model

    init(model: Model, controller: Controller) {
        self.model = model
        self.controller = controller
        self.model.addObserver(self)
    }

    func update() {
        print("View : Update.")
        let data = model.getData()
        print("Data: \(data)")
    }

    func getUserInput() {
        print("View : Enter user input: ")
        //let userInput = readLine()
        let userInput = "Hello, World!" 
        print(userInput)
        controller.handleUserInput(userInput)
    }
}

// Controller
class Controller {
    private var model: Model
    private var view: View?

    init(model: Model) {
        self.model = model
    }

    func handleUserInput(_ userInput: String) {
        print("Controller : Handle user input.")
        model.setData(userInput)
    }

    func setView(_ view: View) {
        self.view = view
    }
}

// Client Code
let model = Model()
let controller = Controller(model: model)
let view = View(model: model, controller: controller)
controller.setView(view)
view.getUserInput()

/*
View : Enter user input: 
Hello, World!
Controller : Handle user input.
Model : Set data.
Model : Notify observers.
View : Update.
Model : Get data.
Data: Hello, World!
*/